File: verify_performance_indexes.py
#!/usr/bin/env python3
"""
CryoProtect v2 - Verify Performance Indexes

This script verifies that the performance indexes have been successfully created
in the CryoProtect Supabase database.

Usage:
    python verify_performance_indexes.py
"""

import os
import sys
import json
import logging
from dotenv import load_dotenv
from supabase import create_client

# Set up logging
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(message)s",
    handlers=[
        logging.FileHandler("verify_performance_indexes.log"),
        logging.StreamHandler()
    ]
)
logger = logging.getLogger(__name__)

# Load environment variables
load_dotenv()

def connect_to_supabase():
    """Connect to Supabase using service role key."""
    supabase_url = os.getenv("SUPABASE_URL")
    supabase_key = os.getenv("SUPABASE_KEY")
    
    if not supabase_url or not supabase_key:
        logger.error("SUPABASE_URL and SUPABASE_KEY must be set in .env file")
        return None
    
    try:
        supabase = create_client(supabase_url, supabase_key)
        logger.info("Connected to Supabase")
        return supabase
    except Exception as e:
        logger.error(f"Error connecting to Supabase: {str(e)}")
        return None

def execute_sql(supabase, sql, description):
    """Execute SQL using Supabase."""
    try:
        logger.info(f"Executing SQL: {description}")
        response = supabase.rpc("exec_sql", {"sql": sql}).execute()
        
        if hasattr(response, 'error') and response.error:
            logger.error(f"Error executing SQL: {response.error}")
            return False, None
        
        return True, response.data
    except Exception as e:
        logger.error(f"Error executing SQL ({description}): {str(e)}")
        return False, None

def verify_indexes(supabase):
    """Verify that the performance indexes exist."""
    # List of indexes to verify
    expected_indexes = [
        {
            "name": "idx_predictions_mixture_property",
            "table": "predictions",
            "columns": ["mixture_id", "property_type_id"],
            "using": None
        },
        {
            "name": "idx_molecule_name_trgm",
            "table": "molecules",
            "columns": ["name"],
            "using": "gin"
        },
        {
            "name": "idx_mixture_name_trgm",
            "table": "mixtures",
            "columns": ["name"],
            "using": "gin"
        }
    ]
    
    # Get all indexes
    sql = """
    SELECT
        i.relname AS index_name,
        t.relname AS table_name,
        array_agg(a.attname ORDER BY k.indnatts) AS column_names,
        am.amname AS index_method
    FROM
        pg_index k
        JOIN pg_class i ON i.oid = k.indexrelid
        JOIN pg_class t ON t.oid = k.indrelid
        JOIN pg_attribute a ON a.attrelid = t.oid AND a.attnum = ANY(k.indkey)
        JOIN pg_am am ON am.oid = i.relam
    WHERE
        t.relkind = 'r'
        AND t.relnamespace = 'public'::regnamespace
        AND i.relnamespace = 'public'::regnamespace
    GROUP BY
        i.relname,
        t.relname,
        am.amname
    ORDER BY
        t.relname,
        i.relname;
    """
    
    success, indexes = execute_sql(supabase, sql, "Get all indexes")
    if not success:
        return False
    
    # Check if each expected index exists
    missing_indexes = []
    for expected in expected_indexes:
        found = False
        for idx in indexes:
            if idx["index_name"] == expected["name"]:
                # Check if it's on the right table
                if idx["table_name"] != expected["table"]:
                    logger.warning(f"Index {expected['name']} is on table {idx['table_name']} instead of {expected['table']}")
                    continue
                
                # Check if it has the right columns
                columns_match = True
                for col in expected["columns"]:
                    if col not in idx["column_names"]:
                        logger.warning(f"Index {expected['name']} is missing column {col}")
                        columns_match = False
                        break
                
                # Check if it uses the right method
                if expected["using"] and idx["index_method"] != expected["using"]:
                    logger.warning(f"Index {expected['name']} uses method {idx['index_method']} instead of {expected['using']}")
                    continue
                
                if columns_match:
                    found = True
                    logger.info(f"Index {expected['name']} exists on table {expected['table']}")
                    break
        
        if not found:
            missing_indexes.append(expected["name"])
            logger.error(f"Index {expected['name']} does not exist on table {expected['table']}")
    
    # Check if pg_trgm extension is installed
    sql = """
    SELECT EXISTS (
        SELECT 1 FROM pg_extension WHERE extname = 'pg_trgm'
    );
    """
    
    success, result = execute_sql(supabase, sql, "Check pg_trgm extension")
    if not success:
        return False
    
    pg_trgm_installed = result[0]["exists"]
    if not pg_trgm_installed:
        logger.error("pg_trgm extension is not installed")
        return False
    
    logger.info("pg_trgm extension is installed")
    
    # Return True if all indexes exist
    if missing_indexes:
        logger.error(f"Missing indexes: {', '.join(missing_indexes)}")
        return False
    
    logger.info("All performance indexes exist")
    return True

def main():
    """Main function to verify performance indexes."""
    print("\n" + "=" * 80)
    print("CryoProtect v2 - Verify Performance Indexes")
    print("=" * 80 + "\n")
    
    # Connect to Supabase
    supabase = connect_to_supabase()
    if not supabase:
        return 1
    
    # Verify indexes
    success = verify_indexes(supabase)
    
    # Print summary
    print("\n" + "=" * 60)
    print("Performance Indexes Verification Summary")
    print("=" * 60)
    
    if success:
        print("\nStatus: PASS")
        print("All performance indexes are correctly installed.")
    else:
        print("\nStatus: FAIL")
        print("One or more performance indexes are missing.")
        print("Please run the apply_performance_indexes script.")
    
    print("\nExpected Indexes:")
    print("  - idx_predictions_mixture_property on predictions(mixture_id, property_type_id)")
    print("  - idx_molecule_name_trgm on molecules using gin(name gin_trgm_ops)")
    print("  - idx_mixture_name_trgm on mixtures using gin(name gin_trgm_ops)")
    
    print("\nFor detailed information, check the log file.")
    print("=" * 60 + "\n")
    
    return 0 if success else 1

if __name__ == "__main__":
    sys.exit(main())

File: verify_performance_indexes.js
#!/usr/bin/env node

/**
 * CryoProtect Analyzer - Performance Indexes Verification Utility
 * 
 * This script verifies that the performance indexes have been successfully created.
 * It requires the Supabase CLI to be installed.
 */

const { execSync } = require('child_process');
const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

// Colors for console output
const colors = {
  reset: '\x1b[0m',
  bright: '\x1b[1m',
  dim: '\x1b[2m',
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  cyan: '\x1b[36m'
};

// Print banner
console.log(`
${colors.cyan}${colors.bright}=================================================
 CryoProtect Analyzer - Performance Indexes Verification
=================================================
${colors.reset}
This utility will verify that the performance indexes
have been successfully created in your Supabase database.
`);

// Check if Supabase CLI is installed
try {
  execSync('supabase --version', { stdio: 'ignore' });
  console.log(`${colors.green}✓ Supabase CLI is installed${colors.reset}`);
} catch (error) {
  console.error(`
${colors.red}✗ Supabase CLI is not installed or not in PATH${colors.reset}

Please install the Supabase CLI first:
  npm install -g supabase
  
Then run this script again.
`);
  process.exit(1);
}

// Ask for Supabase project reference
rl.question(`
${colors.yellow}Enter your Supabase project reference:${colors.reset} `, (projectRef) => {
  if (!projectRef.trim()) {
    console.error(`
${colors.red}✗ Project reference is required${colors.reset}

You can find your project reference in the Supabase dashboard:
1. Go to https://app.supabase.io/
2. Select your project
3. Go to Project Settings > General
4. Copy the "Reference ID"
`);
    rl.close();
    process.exit(1);
  }

  console.log(`
${colors.yellow}Linking to Supabase project: ${projectRef}${colors.reset}
`);

  try {
    // Link to Supabase project
    execSync(`supabase link --project-ref ${projectRef}`, { stdio: 'inherit' });
    
    console.log(`
${colors.green}✓ Successfully linked to Supabase project${colors.reset}
    
${colors.yellow}Verifying performance indexes...${colors.reset}
`);

    // SQL queries to check for the existence of each index
    const verificationQueries = [
      {
        name: "Composite index on predictions(mixture_id, property_type_id)",
        query: "SELECT EXISTS (SELECT 1 FROM pg_indexes WHERE indexname = 'idx_predictions_mixture_property')"
      },
      {
        name: "Text search index on molecule names",
        query: "SELECT EXISTS (SELECT 1 FROM pg_indexes WHERE indexname = 'idx_molecule_name_trgm')"
      },
      {
        name: "Text search index on mixture names",
        query: "SELECT EXISTS (SELECT 1 FROM pg_indexes WHERE indexname = 'idx_mixture_name_trgm')"
      },
      {
        name: "pg_trgm extension",
        query: "SELECT EXISTS (SELECT 1 FROM pg_extension WHERE extname = 'pg_trgm')"
      }
    ];

    // Create a temporary file with the verification queries
    const fs = require('fs');
    const path = require('path');
    const tempFile = path.join(__dirname, 'temp_verification.sql');
    
    let allIndexesExist = true;
    
    for (const query of verificationQueries) {
      // Write the query to a temporary file
      fs.writeFileSync(tempFile, query.query);
      
      // Execute the query
      const result = execSync(`supabase db execute --file ${tempFile}`, { encoding: 'utf8' });
      
      // Check if the index exists
      const exists = result.trim().toLowerCase().includes('true');
      
      if (exists) {
        console.log(`${colors.green}✓ ${query.name} exists${colors.reset}`);
      } else {
        console.log(`${colors.red}✗ ${query.name} does not exist${colors.reset}`);
        allIndexesExist = false;
      }
    }
    
    // Clean up the temporary file
    fs.unlinkSync(tempFile);
    
    if (allIndexesExist) {
      console.log(`
${colors.green}${colors.bright}✓ All performance indexes have been successfully created!${colors.reset}

Your CryoProtect Analyzer database has been optimized with the following indexes:
- Composite index on predictions(mixture_id, property_type_id)
- Text search index on molecule names using pg_trgm
- Text search index on mixture names using pg_trgm

These indexes should significantly improve query performance.
`);
    } else {
      console.log(`
${colors.yellow}${colors.bright}⚠ Some performance indexes are missing.${colors.reset}

Please run the apply_performance_indexes script to create all the required indexes.
`);
    }
  } catch (error) {
    console.error(`
${colors.red}✗ An error occurred while verifying the indexes${colors.reset}

Error details:
${error.message}

Please check your project reference and try again.
`);
  }
  
  rl.close();
});